File: CynapCRM.Services.InventoryAPI/Service/StockMovementService.cs
using AutoMapper;
using CynapCRM.Services.InventoryAPI.Data;
using CynapCRM.Services.InventoryAPI.Models;
using CynapCRM.Services.InventoryAPI.Models.Dto;
using CynapCRM.Services.InventoryAPI.Service.IService;
using Microsoft.EntityFrameworkCore;

namespace CynapCRM.Services.InventoryAPI.Service
{
    public class StockMovementService : IStockMovementService
    {
        private readonly AppDbContext _db;
        private readonly IMapper _mapper;

        public StockMovementService(AppDbContext db, IMapper mapper)
        {
            _db = db;
            _mapper = mapper;
        }
        public async Task<bool> DecrementStockAsync(int idStock, int qte)
        {
            if (qte <= 0)
            {
                return false;
            }

            var stock = await _db.StocksDelegues
                .FirstOrDefaultAsync(s => s.Id_stock == idStock && !s.IsDeleted);
            if (stock == null || stock.QteDisponible < qte)
            {
                return false;
            }
            stock.QteDisponible -= qte;
            _db.StockMovements.Add(new StockMovement
            {
                Id_Stock = idStock,
                Quantite = -qte,
                DateMovement = DateTime.UtcNow,
                TypeMovement = "Decrement"
            });

            await _db.SaveChangesAsync();
            return true;

        }
        public async Task<bool> IncrementStockAsync(int idStock, int qte)
        {
            if (qte <= 0)
            {
                return false;
            }

            var stock = await _db.StocksDelegues
                            .FirstOrDefaultAsync(s => s.Id_stock == idStock && !s.IsDeleted);
            if (stock == null)
            {
                return false;
            }
            stock.QteDisponible += qte;
            _db.StockMovements.Add(new StockMovement
            {
                Id_Stock = idStock,
                Quantite = qte,
                DateMovement = DateTime.UtcNow,
                TypeMovement = "Increment"
            });
            await _db.SaveChangesAsync();
            return true;
        }
        public async Task<bool> TransferStockAsync(int idStockSource, int idStockDestination, int qte)
        {
            if (qte <= 0) 
            {
                return false;
            }

            var source = await _db.StocksDelegues
                            .FirstOrDefaultAsync(s => s.Id_stock == idStockSource && !s.IsDeleted);

            var destination = await _db.StocksDelegues
                .FirstOrDefaultAsync(s => s.Id_stock == idStockDestination && !s.IsDeleted);

            if (source == null || destination == null || source.QteDisponible < qte)
            {
                return false;
            }
            
            using var transaction = await _db.Database.BeginTransactionAsync();
            try
            {
                source.QteDisponible -= qte;
                destination.QteDisponible += qte;

                _db.StockMovements.Add(new StockMovement
                {
                    Id_Stock = idStockSource,
                    TypeMovement = "Transfer-Out",
                    Quantite = qte,
                    DateMovement = DateTime.UtcNow,
                    Description = $"Transfert vers stock {idStockDestination}"
                });

                _db.StockMovements.Add(new StockMovement
                {
                    Id_Stock = idStockDestination,
                    TypeMovement = "Transfer-In",
                    Quantite = qte,
                    DateMovement = DateTime.UtcNow,
                    Description = $"Transfert depuis stock {idStockSource}"
                });

                await _db.SaveChangesAsync();
                await transaction.CommitAsync();
                return true;
            }
            catch
            {
                await transaction.RollbackAsync();
                return false;
            }
        }
        public async Task<IEnumerable<StockMovementDto>> GetStockMovementsAsync(int idStock)
        {

            return await _db.StockMovements
                            .AsNoTracking()
                            .Where(m => m.Id_Stock == idStock)
                            .OrderByDescending(m => m.DateMovement)
                            .Select(m => new StockMovementDto
                            {
                                Id_Movement = m.Id_Movement,
                                Id_Stock = m.Id_Stock,
                                Quantite = m.Quantite,
                                TypeMovement = m.TypeMovement,
                                DateMovement = m.DateMovement,
                                Description = m.Description
                            })
                            .ToListAsync();
        }
    }
}

File: Cynapharm/src/app/features/inventory/movements/services/stock-movement.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { ApiService } from '../../../../core/services/api.service';

export interface StockMovementDto {
  id_Movement?: number;
  id_Stock:     number;
  quantite:     number;
  typeMovement: string;
  dateMovement?: string;
  description?: string;
}

@Injectable({ providedIn: 'root' })
export class StockMovementService {
  private readonly base = '/inventory/stock-movements';
  constructor(private api: ApiService) {}
  private u<T>(r: any): T { return r?.Result ?? r?.result ?? r; }

  getMovements(idStock: number): Observable<StockMovementDto[]> {
    return this.api.get<any>(`${this.base}/${idStock}`).pipe(map(r => this.u<StockMovementDto[]>(r) ?? []));
  }
  decrement(idStock: number, qte: number): Observable<boolean> {
    return this.api.post<any>(`${this.base}/decrement?idStock=${idStock}&qte=${qte}`, {}).pipe(map(r => this.u<boolean>(r)));
  }
  increment(idStock: number, qte: number): Observable<boolean> {
    return this.api.post<any>(`${this.base}/increment?idStock=${idStock}&qte=${qte}`, {}).pipe(map(r => this.u<boolean>(r)));
  }
  transfer(idSource: number, idDest: number, qte: number): Observable<boolean> {
    return this.api.post<any>(`${this.base}/transfer?idStockSource=${idSource}&idStockDestination=${idDest}&qte=${qte}`, {}).pipe(map(r => this.u<boolean>(r)));
  }
}
